Surfer reader feature

jessepisel opened the pull request:
ported the surfer reader from fatiando and changed the output to a xarray Data Array. Ran tests and formatted locally. 

I ported the surfer reader feature from fatiando and changed the output from a dict to an xarray data array. Will you look through it to make sure it's what you had in mind for this feature. Do you want me to add examples to the docstring as well? Right now it just has the docstring from fatiando modified as appropriate.

Fixes #38

**Reminders**

- [x] Run `make format` and `make check` to make sure the code follows the style guide.
- [x] Add tests for new features or tests that would have caught the bug that you're fixing.
- [x] Add new public functions/methods/classes to `doc/api/index.rst` and `verde/__init__.py`.
- [x] Write detailed docstrings for all functions/classes/methods.
- [ ] If adding new functionality, add an example to the docstring, gallery, and/or tutorials.

leouieda commented on verde/utils.py:
```diff
@@ -281,3 +282,64 @@ def grid_to_table(grid):
+    output being a dictionary, in verde the output is a `xarray.DataArray`
+    that contains the easting and northing coordinates, and data values.
+
+    Parameters:
```
```suggestion
    Parameters
    ----------
```

leouieda commented on verde/utils.py:
```diff
@@ -281,3 +282,64 @@ def grid_to_table(grid):
+    that contains the easting and northing coordinates, and data values.
+
+    Parameters:
+
```
```suggestion
```

leouieda commented on verde/utils.py:
```diff
@@ -281,3 +282,64 @@ def grid_to_table(grid):
+        The type of variable used for the data. Default is numpy.float64. Use
+        numpy.float32 if the data are large and precision is not an issue.
+
+    Returns:
```
```suggestion
    Returns
    -------
```

leouieda commented on verde/utils.py:
```diff
@@ -281,3 +282,64 @@ def grid_to_table(grid):
+        numpy.float32 if the data are large and precision is not an issue.
+
+    Returns:
+
```
```suggestion
```

leouieda commented on verde/utils.py:
```diff
@@ -281,3 +282,64 @@ def grid_to_table(grid):
+
+    * fname : str
+        Name of the Surfer grid file
+    * dtype : numpy dtype object or string
```
```suggestion
    dtype : str or numpy dtype
```

leouieda commented on verde/utils.py:
```diff
@@ -281,3 +282,64 @@ def grid_to_table(grid):
+
+    Parameters:
+
+    * fname : str
```
```suggestion
    fname : str
```

leouieda commented on verde/utils.py:
```diff
@@ -281,3 +282,64 @@ def grid_to_table(grid):
+    Parameters:
+
+    * fname : str
+        Name of the Surfer grid file
```
```suggestion
        Name or path of the Surfer grid file.
```

leouieda commented on verde/utils.py:
```diff
@@ -281,3 +282,64 @@ def grid_to_table(grid):
+    * fname : str
+        Name of the Surfer grid file
+    * dtype : numpy dtype object or string
+        The type of variable used for the data. Default is numpy.float64. Use
```
```suggestion
        The type of variable used for the data. Default is ``float64``. Use
```

leouieda commented on verde/utils.py:
```diff
@@ -281,3 +282,64 @@ def grid_to_table(grid):
+        Name of the Surfer grid file
+    * dtype : numpy dtype object or string
+        The type of variable used for the data. Default is numpy.float64. Use
+        numpy.float32 if the data are large and precision is not an issue.
```
```suggestion
        ``float32`` if the data are large and precision is not an issue.
```

leouieda commented on verde/utils.py:
```diff
@@ -281,3 +282,64 @@ def grid_to_table(grid):
+
+    Returns:
+
+    * data : :class: `xarray.DataArray`
```
```suggestion
    data : :class:`xarray.DataArray`
```

leouieda commented on verde/utils.py:
```diff
@@ -281,3 +282,64 @@ def grid_to_table(grid):
+        assert np.allclose(dmin, field.min()), err_msg.format('Min', dmin,
+                                                              field.min())
+        assert np.allclose(dmax, field.max()), err_msg.format('Max', dmax,
+                                                              field.max())
```
Would you mind converting the `assert` statements to `if ...: raise IOError(...)` blocks instead?

jessepisel replied:
Changed both `assert `statements to `if... raise IOError`

leouieda commented on verde/utils.py:
```diff
@@ -281,3 +282,64 @@ def grid_to_table(grid):
+                                                              field.max())
+        data = xr.DataArray(field.reshape(ydims, xdims), coords=(np.linspace(
+        ymin, ymax, ydims), np.linspace(xmin, xmax, xdims)), dims=['northing',
+        'easting'])
```
It would be great to fill the `attrs` attribute of the `DataArray` with some provenance information about the file. Including the `DSAA` and original file name would be a good start. The boundaries are probably not necessary since they can be calculated from the coordinates easily.

jessepisel replied:
Great suggestion, I added some additional attributes to the `DataArray`

leouieda commented on verde/utils.py:
```diff
@@ -281,3 +282,64 @@ def grid_to_table(grid):
+
+    This feature was ported from Fatiando, but instead of
+    output being a dictionary, in verde the output is a `xarray.DataArray`
+    that contains the easting and northing coordinates, and data values.
```
The mention of Fatiando isn't necessary here. But it is good to describe what is returned by the function and some other details (like the masking of NaNs and inclusion of metadata in `attrs`).

leouieda commented on verde/utils.py:
```diff
@@ -281,3 +282,64 @@ def grid_to_table(grid):
+        field = np.fromiter((float(s)
+                             for line in input_file
+                             for s in line.split()),
+                            dtype=dtype)
```
This call should be replaced with a `numpy.loadtxt`. Then we don't need to reshape the array before passing it to `DataArray`. It would also let us check the shape against `ydims`and `xdims`.

jessepisel replied:
This one I am having trouble with. When I read in .grd files `numpy.loadtxt` is having issues with line endings. I will keep working on this.

leouieda replied:
I vaguely remember having similar issues. Let me know if you need any help with this.

jessepisel replied:
I haven't been able to figure out an easy way to parse through a `.grd` file with `np.loadtxt` because of the file formatting. It looks like it uses a `\n` newline character every 10 entries and then again at the end of each row with ` \n` with a space. It's fairly easy to count the number of ` \n` entries and get the number of rows, but there has to be a way to do it all in one pass with `np.loadtxt` or `np.genfromtxt`. Input is very welcome

leouieda replied:
Right, I forgot about those blank lines that the surfer grid has. Probably best to not use `loadtxt` then. But you could still add a check that the number of elements read is consistent with `ydims*xdims`.

jessepisel replied:
I added a check that the dimensions from the header match the data

leouieda commented:
I'm going to fix the merge conflicts here so please `git pull` the new commit before making any changes.

leouieda commented:
@santisoler could you also take a look and see if you have anything to add or some suggestions?

santisoler commented:
Hi @jessepisel, nice work porting it. I think this kind of functions makes a lot easier the migration of our previous workflow on other software (like Surfer in this case) to fatiando. Also, it really helps to integrate ourselves among research groups that still works with this kind of data files.

I'll leave some comments on the code. Don't feel discouraged by the proposed changes, they are mainly on styling, which is kind of subjective thing that changes in time, so nothing to worry about.

santisoler commented on verde/utils.py:
```diff
@@ -254,6 +255,67 @@ def grid_to_table(grid):
+        field = np.fromiter((float(s)
+                             for line in input_file
+                             for s in line.split()),
+                            dtype=dtype)
```
What about renaming the `x` and `y` coordinates to something like `northing` and `easting`? I don't know if Surfer grids are also used with other kind of data than 2D horizontal coordinates (like a 2D profile where one of the coordinates will be the vertical dimension). In any case would be nice to explicit which coordinate corresponds to the northing and which one to the easting.

jessepisel replied:
I like your suggestion, are you thinking `xmin` to `eastmin` or something like that?

leouieda replied:
Or `east`, `west`, `south`, `north` to make it extra clear.

santisoler commented on verde/utils.py:
```diff
@@ -254,6 +255,67 @@ def grid_to_table(grid):
+                                                              field.max())
+        data = xr.DataArray(field.reshape(ydims, xdims), coords=(np.linspace(
+        ymin, ymax, ydims), np.linspace(xmin, xmax, xdims)), dims=['northing',
+        'easting'])
```
Also will be nice to make the `DataArray` definition more readable, maybe something like [this](https://github.com/fatiando/harmonica/blob/master/harmonica/io.py#L36-L56).

In this case, it could be something like this:
```python
dims = ["northing", "easting"]
coords = {
    "northing": np.linspace(ymin, ymax, ydims),
    "easting": np.linspace(xmin, xmax, xdims)
}
attrs = dict(...)
field = field.reshape(ydims, xdims)
data = xr.DataArray(field, coords=coords, dims=dims, attrs=attrs)
```

jessepisel replied:
Yours looks much better and easier to read. Thank you for making this change, I have added it to the updated commit

jessepisel commented:
@leouieda @santisoler Thank you both for the reviews, they were very helpful, I am glad to see I am on the right track! I implemented the easy formatting changes and added other suggested changes. I am going to start working on some tests for this and continue working on the `numpy.loadtxt` issues I have been having. Hopefully I will have a PR in the next few days.

leouieda commented on verde/utils.py:
```diff
@@ -254,6 +255,72 @@ def grid_to_table(grid):
+
+    http://www.goldensoftware.com/products/surfer
+
+    This feature opens a Surfer grid file, masks any NaNs in the data,
```
```suggestion
    This function opens a Surfer grid file, masks any NaNs in the data,
```

leouieda commented on verde/utils.py:
```diff
@@ -254,6 +255,72 @@ def grid_to_table(grid):
+    http://www.goldensoftware.com/products/surfer
+
+    This feature opens a Surfer grid file, masks any NaNs in the data,
+    and outputs a `xarray.DataArray` that contains easting and northing
```
```suggestion
    and outputs a :class:`xarray.DataArray` that contains easting and northing
```

leouieda replied:
This will automatically create a link to the documentation of the DataArray (which I find really handy).

leouieda commented on verde/utils.py:
```diff
@@ -254,6 +255,72 @@ def grid_to_table(grid):
+        if np.any(nans):
+            field = np.ma.masked_where(nans, field)
+        err_msg = "{} of data ({}) doesn't match one from file ({})."
+        if np.allclose(dmin, field.min()) == False:
```
```suggestion
        if not np.allclose(dmin, field.min()):
```

This should fix some of the pylint complaints. It's usually not good to compare to booleans in `if` statements.

jessepisel replied:
Rad, I wasn't positive that was the best way to do that.

leouieda commented on verde/utils.py:
```diff
@@ -254,6 +255,72 @@ def grid_to_table(grid):
+    # zMin zMax       Z min max
+    # z11 z21 z31 ... List of Z values
+    with open(fname) as input_file:
+        # DSAA is a Surfer ASCII GRD ID (discard it for now)
```
```suggestion
```

leouieda replied:
Oh the dangers of outdated comments :wink:

leouieda commented on verde/utils.py:
```diff
@@ -254,6 +255,72 @@ def grid_to_table(grid):
+    # z11 z21 z31 ... List of Z values
+    with open(fname) as input_file:
+        # DSAA is a Surfer ASCII GRD ID (discard it for now)
+        grd_id = input_file.readline()
```
```suggestion
        grd_id = input_file.readline().strip()
```

To remove any extra spaces or newlines.

jessepisel replied:
that's the call I was looking for but couldn't remember...

leouieda commented on verde/utils.py:
```diff
@@ -254,6 +255,72 @@ def grid_to_table(grid):
+            "easting": np.linspace(xmin, xmax, xdims)
+        }
+        field = field.reshape(ydims, xdims)
+        attrs = dict(file=fname, grid_id=grd_id[:-1])
```
```suggestion
        attrs = {"file": fname, "DSAA grid ID": grid_id}
```

leouieda commented:
Also, it's good to run `make format` before committing to avoid errors in Travis CI due to code formatting. `make check` and `make lint` also give good warnings (usually) about possible flaws in the code.

leouieda commented on verde/utils.py:
```diff
@@ -254,6 +255,73 @@ def grid_to_table(grid):
+        if np.any(nans):
+            field = np.ma.masked_where(nans, field)
+        err_msg = "{} of data ({}) doesn't match one from file ({})."
+        if not np.isclose(len(field), ydims * xdims):
```
```suggestion
        if field.size != ydims * xdims:
```

The `np.isclose` function is used to test floating point numbers. Since all are integers `!=` will work better. With a numpy array, it's best to check `.size` or `.shape` since `len` of a 2D array will be number of rows instead of the number of elements. I've had some bugs creep up in the past because of that.

jessepisel replied:
Right on, thank you for the feedback. I have commited the changes, and will update the tests that I am working on

jessepisel commented:
I have some tests for the header parser and grid parser portions of the function in my forked version/surfer branch of verde. Do you want them in a new PR for them or how would you like them for review?

santisoler commented:
Hi @jessepisel. Any PR that wants to be merged into `master` must have 100% test coverage. So when any new enhancement is incorporated into the PR, it should be accompanied with its corresponding tests functions. This means that every test function for your new Surfer grid reader function must be included on this PR.

By the way, I've seen you added pieces of software belonging to the Surfer reader function into the test functions. Although this may be considered a good way to test the function itself (because this pieces of software are the same), it's a source of problems for the future. For example, if we make any change in the original function that involuntary includes a bug, the test function will not fail because it's testing old code, clearing the path for an undetected bug.

So the idea is to call the functions that wants to be tested by importing them first (as you would use them on your notebooks or scripts).  See the [`load_icgem`](https://github.com/fatiando/harmonica/blob/master/harmonica/tests/test_icgem.py) function from Harmonica and its corresponding [test functions](https://github.com/fatiando/harmonica/blob/master/harmonica/tests/test_icgem.py) to get an idea of what I'm saying.

What @leouieda meant by separating the function is to create two functions in `verde/utils.py`: one called `load_surfer()` and the other one could be `_read_surfer_header()`. The idea is to leave the header reading task to the latter, while the first one only calls it to get the header information.

Something like this could be an example:
```python
def load_surfer(fname, dtype="float64"):
    """
    Read data from a Surfer ASCII grid file.
    ...
    """
    with open(fname) as input_file:
        grid_id, ydims, xdims, south, north, west, east, dmin, dmax = _read_surfer_header(fname)
        field = np.fromiter(
            (float(s) for line in input_file for s in line.split()), dtype=dtype
        )
    ...
    return data


def _read_surfer_header(input_file):
    # DSAA is a Surfer ASCII GRD ID
    grd_id = input_file.readline().strip()
    # Read the number of columns (ny) and rows (nx)
    ydims, xdims = [int(s) for s in input_file.readline().split()]
    # Our x points North, so the first thing we read is y, not x.
    south, north = [float(s) for s in input_file.readline().split()]
    west, east = [float(s) for s in input_file.readline().split()]
    dmin, dmax = [float(s) for s in input_file.readline().split()]
    return grid_id, ydims, xdims, south, north, west, east, dmin, dmax
```

The same goes for the validation data lines, for example:
```python
def _check_surfer_integrity(...):
    err_msg = "{} of data ({}) doesn't match one from file ({})."
    if field.size != ydims * xdims:
        raise IOError(err_msg.format("Dimensions", (ydims * xdims), len(field)))
    if not np.allclose(dmin, field.min()):
        raise IOError(err_msg.format("Min", dmin, field.min()))
    if not np.allclose(dmax, field.max()):
        raise IOError(err_msg.format("Max", dmax, field.max()))
```

We put an underscore before the function name to make clear that it's a private function, i.e. a function that is not supposed to be called by users and it's only used buy functions inside the module to perform internal tasks.

Hope my comment remove any doubts you have. Feel free to ask for more help without any problem!

leouieda commented:
@santisoler beat me to it :wink: 

The idea is that "adding tests later" tends to never happen. I know this from personal experience and it's the reason the testing situation in Fatiando was so inconsistent. So with Verde I decided to make this a priority from the start. If you need help setting this up, just let us know.

jessepisel commented:
Thanks for the detailed explanation @santisoler and @leouieda I will keep plugging away at the tests and see what I can come up with!

leouieda commented:
@jessepisel no problem, take your time. Don't hesitate to call on us if you want any help (we can write some code and push to this branch as well). Testing is hard and it takes time to get used to it.

jessepisel commented:
I updated the function and split it into several sub-functions. I need to reformat it with the new line lengths discussion that I saw this past week. I'll work on some tests this week and have them submitted by the weekend.

leouieda commented:
Thanks for the update @jessepisel! Let me know when you're ready and I can jump in to review the code.

jessepisel commented:
I think this is ready for a code review if you have the time to look through it.

leouieda commented:
Hi @jessepisel, I'm having a look at your code right now. I think I'll just make some commits with a few fixes right to this PR so we don't delay this much more. I want to have it in the next release.

jessepisel commented:
That sounds good to me. Let me know what I can do to help.

leouieda commented:
@jessepisel OK, I refactored the code to be a bit shorter and more concise. I also added a few test options (with/without nans, from StringIO or from file). The function now can take an open file or a file name. Please have a look and see what you think.

When coding for one of our projects, make sure you use `make check`, `make lint`, and `make format` often to catch some possible errors. I run these all the time and it really helps to find things like unused variables and imports, invalid syntax, etc.

leouieda commented:
@santisoler would you mind reviewing this? If you're happy with it feel free to squash and merge. When doing that, please edit the commit message to summarize what this PR is contributing.

jessepisel commented:
I like your changes, it's easier to read with the formatting and it is to the point. Thanks for making those changes to it!

santisoler commented:
> @santisoler would you mind reviewing this? If you're happy with it feel free to squash and merge. When doing that, please edit the commit message to summarize what this PR is contributing.

Ok @leouieda. I'm quite busy today, but I'll leave it for next week.

Nice work @jessepisel! I always learn a lot when someone else refactor my codes. It's a great opportunity to acquire more knowledge on coding and software development!

leouieda commented:
:tada: :tada: @jessepisel thanks for the contribution! :tada: :tada: 

I'll cut a new release as soon as #181 is merged so this will make it into 1.2.0.

Thanks @santisoler for reviewing and merging!

jessepisel commented:
Thanks for all the help @leouieda and @santisoler I learned a lot from this one. I appreciate all the help and guidance on putting this together!

leouieda commented:
:+1: I hope to see new ones in the future :wink:
